### PR TITLE
Nested creation

### PR DESCRIPTION
**Pull request by mickhansen**

### CODE REVIEW

File: lib/instance.js [hunk 1]
```diff
@@ -638,6 +662,36 @@ module.exports = (function() {
...
+              if (!instances.length) return Promise.resolve();
+
+              // Instances will be updated in place so we can safely treat HasOne like a HasMany
+              return Promise.map(instances, function (instance) {
+                if (include.association instanceof BelongsToMany) {
+                  return instance.save({transaction: options.transaction, logging: console.log}).then(function () {
+                    var values = {};
+                    values[include.association.foreignKey] = self.get(self.Model.primaryKeyAttribute, {raw: true});
+                    values[include.association.otherKey] = instance.get(instance.Model.primaryKeyAttribute, {raw: true});
+                    return include.association.throughModel.create(values, {transaction: options.transaction, logging: console.log});
```
Comment (mickhansen) [inline lib/instance.js]: whoops

**Comment by mbroadst** [comment]
LGTM

**Comment by pilsy** [comment]
nice :+1:

**Comment by troyastorino** [comment]
What are the thoughts for further work on this? I would love to see it extended to work for `update`. 

@mickhansen I saw you said in #3169 that `update` would be way harder. Would that be because it would touch code in a lot more places, or because there are more situations you'd have to cover?

**Comment by mickhansen** [comment]
@troyastorino There are a ton more cases for update yes. Especially a case like UUID primary keys, how do you know whether it's a create or an existing - and a lot more cases like that ;p

**Comment by troyastorino** [comment]
Ah ya, makes sense. 

Regardless, I'd love to see this, and hope I might be able to find some time to work on it at some point soon. Maybe a good first step would be for me to write test cases for all the situations I can think of, and then you and other maintainers could review and point out all the ones I'm missing? :)

**Comment by mickhansen** [comment]
@troyastorino Test cases are a great help, especially if you can seperate them into a few PRs.
We could also definitely start out by only supporting AI/SERIAL primary keys in the beginning.

**Comment by fixe** [comment]
This PR broke support for building models hydrated with an already-inserted relation. For example, this used to work:

```js
var user = <a pre-existing User model instance>

Product.create({
  user: user,
  UserId: user.id
}, { include: [User] }).then(function(product) {
   // `product.user` would be set to the instance we provided above.
});
```

Since this PR, when *INSERT*ing the product, sequelize also attempts to insert the *User*, which will fail due to the unique constraint. However, the PR introduces a useful feature -- maybe it's possible to support both use-cases?

**Comment by mickhansen** [comment]
@fixe Hmm yeah that's a case we didn't exactly consider.
Can you give this a shot:

```js
Product.create({
  UserId: user.id
}, { include: [User] }).then(function(product) {
  product.set('user', user);
   // `product.user` would be set to the instance we provided above.
});
```

**Comment by fixe** [comment]
@mickhansen that's our current solution but it feels like sequelize should be handling this case.

**Comment by mickhansen** [comment]
@fixe That would be the ideal :) There are quite a few edge cases we need to consider, raw data vs instances, AI PKs vs UUID pks (for instance), for non AI PK's we have no idea if the content needs to be created, updated or no-op'ed (unless wrapped in an instance which has isNewRecord)

**Comment by grabbou** [comment]
Is there any way to use this just to set many-to-many relations during a creation of the object?
In my case
```js
Product.create({
  Users: [1, 2, 3, 4, 5]
}, { include: [User] }).then(function(product) {
  
});
```
Sequelize throws constraint error. I would just expect users to be set and returned as a `product.users`. Is there any relatively fast and efficient way to accomplish that?

**Comment by janmeier** [comment]
@grabbou As noted above, this only works for new instances, not for existing ones.

However, you can pass an array of ids to setUsers:

```js
Product.create().then(function(product) {
  product.setUsers([1, 2, 3, 4, 5]);
});
```

**Comment by grabbou** [comment]
Yeah, I know :) The only thing I am trying to solve is to avoid extra call to get the result of `set` operation. In my case, I am having a REST endpoint where user sends a title of a `Product` and passes `ids` of Categories. For now, I've managed to do that with `create(include: Categories)`, `setCategories(ids)`, `reload`, just because I needed to return full output to a client (with all categories filled in). (Un)fortunately, the product of `setCategories` is a ThroughModel.

**Comment by mshahriarinia** [comment]
Great feature! +1 for update.

**Comment by bpmckee** [comment]
How would you handle an insert that's 3 levels deep?  That seems really hard to do.  For exmaple,
```
football = {
  name: 'NFL',
  Teams: [
    {
      name: 'Minnesota Vikings',
      GameDates: [ { date: '9/7/2015' }, { date: '9/22/2015' }, ... ]
    },
    {
      name: 'Seattle Seahawks,
      GameDates: [ { date: '9/10/2015' }, { date: '9/22/2015' }, ... ]
    },
    ...
  ]
};
```

Ideally I'd be able to do the insert like this:
```
return Sport.create(football, {
    include: [{
      model: Team,
      as: 'Teams',
      include: [{
        model: GameDate,
        as: 'GameDates'
      }]
    }]
  });
```

This would throw a notNull violation because `sportId` on the `Team` model would not be defined yet and it'd break there.

I don't even know how I'd have to do this create.  Probably something like this.
```
let response;
return sequelize.transaction(t => {
  return Sport.create(football, { transaction: t })
    .then(sport => {
      response = sport;
      const teamPromises = [];
      football.Teams.forEach(team => {
        team.sportId = sport.id;
        teamPromises = Team.Create(team, { transaction: t });
      });
      return Promise.all(teamPromises);
    })
    .then(teamResults => {
      response.Teams = teamResults;
      const gameDatePromises = [];
      teamResults.forEach(tr => {
        // figure out how to map tr to Team in the original football object
        team.GameDates.forEach(gameDt => {
          gameDt.teamId = teamResult.id; // teamResult was also mapped somehow.
        });
        gameDatePromises.push(GameDate.bulkCreate(team.GameDates, { transaction: t }));
      });
      return Promise.all(gameDatePromises);
    })
    .then(gameDateResults => {
      // somehow map these results back into the response object we've been building.
      return new Promise(resolve => resolve(response));
    });
});
```
A lot more complex and there's some weird mapping steps that I haven't figured out yet.  Please tell me there's an easier way to do this @mickhansen

**Comment by mickhansen** [comment]
@bpmckee That's a known issue with creating a belongsTo with nested, you'll need to add `foreignKey: Math.random().toString()` to the parent level object as a work around.

**Comment by hendrul** [comment]
+1 For update! What has been done so far? where may I start to help?

**Comment by hendrul** [comment]
@bpmckee I test that three levels deep creation problem you mention and it works, is it I'm missing something?
```javascript
var Sequelize = require('sequelize');
var sq = new Sequelize(null,null,null, { 
	dialect: 'sqlite',
	define: {timestamps: false, freezeTableNames: true}
});

var Sport = sq.define('Sport', {
   name: Sequelize.STRING
});

var Team = sq.define('Team', {
	name: Sequelize.STRING
});
var GameDate = sq.define('GameDate', {
	date: Sequelize.DATE
});

Sport.hasMany(Team, {
	foreignKey: {
		allowNull: false
	}
});
Team.hasMany(GameDate, {
	foreignKey: {
		allowNull: false
	}
});

sq.sync({force: true}).then(function(){
	
	var football = {
	  name: 'NFL',
	  Teams: [
	    {
	      name: 'Minnesota Vikings',
	      GameDates: [ { date: '9/7/2015' }, { date: '9/22/2015' }]
	    },
	    {
	      name: 'Seattle Seahawks',
	      GameDates: [ { date: '9/10/2015' }, { date: '9/22/2015' }]
	    }
	  ]
	};

	Sport.create(football, {
	   include: [{
	     model: Team,
	     as: 'Teams',
	     include: [{
	       model: GameDate,
	       as: 'GameDates'
	     }]
	   }]
	}).then(function(instance){
       console.log("Nested creation works well");
	}).catch(function(err){
	   console.error("You're right, there is a problem");
	});

})
```

**Comment by bpmckee** [comment]
@hendrul interesting.

I'm not exactly sure what my differences were but the one thing I noticed was I setup the associations in both directions.  Maybe I shouldn't have done that.

My setup would have looked like this:
```
Sport.hasMany(Team, {
    foreignKey: {
        allowNull: false
    }
});
Team.belongsTo(Sport, {
    foreignKey: 'sport_id'
});
Team.hasMany(GameDate, {
    foreignKey: {
        allowNull: false
    }
});
GameDate.belongsTo(Team, {
    foreignKey: 'team_id'
});
Team.hasOne(Coach, {
    foreignKey: { allowNull: false }
});
Coach.belongsTo(Team, {
    foreignKey: 'team_id'
});
```

**Comment by hendrul** [comment]
This doesn't supports association scopes

**Comment by Ajaxy** [comment]
Is it still impossible to add associations to existing models during create?
(example https://github.com/sequelize/sequelize/pull/3386#issuecomment-97711599)

**Comment by jasonbodily** [comment]
So where are we with associations on build? Currently I have to do the following so as to have it all save together:

```
let person = Person.build()

/* update person attributes */

let person_obj  = person.toJSON();
person_obj.cars = cars; // Association


Person.create(person_obj, {
  include: [{ model: Cars, as: 'cars'}]
}) 
```

Is there a more abbreviated way of doing this?

**Comment by hendrul** [comment]
Right now I need nested creation with existing rows :). Anybody interested in this case?, this PR would be great

**Comment by hendrul** [comment]
I ask, isn't the presence of the pk attribute enough to say that we want to associate instead of making an insert? or it's not a good idea for some reason?

**Comment by mickhansen** [comment]
@hendrul Only in cases where you know that IDs are generated by the database

**Comment by hendrul** [comment]
What if we add support for some cases, this pks generated by db is pretty common, don't you think?

**Comment by mickhansen** [comment]
It's common but these days i normally suggest people use UUIDs instead.
As for adding support, i'm not an active maintainer anymore, just adding why it's not as simple as looking for the PK

**Pull request merged**